prov/util/src/util_coll.c

```diff
@@ -0,0 +1,1042 @@
+		}
+
+#define UTIL_COLL_MIN(dst,src)	if ((dst) > (src)) (dst) = (src)
+#define UTIL_COLL_MAX(dst,src)	if ((dst) < (src)) (dst) = (src)
+#define UTIL_COLL_SUM(dst,src)	(dst) += (src)
+#define UTIL_COLL_PROD(dst,src)	(dst) *= (src)
+#define UTIL_COLL_LOR(dst,src)	(dst) = (dst) || (src)
+#define UTIL_COLL_LAND(dst,src)	(dst) = (dst) && (src)
+#define UTIL_COLL_BOR(dst,src)	(dst) |= (src)
+#define UTIL_COLL_BAND(dst,src)	(dst) &= (src)
+#define UTIL_COLL_LXOR(dst,src)	(dst) = ((dst) && !(src)) || (!(dst) && (src))
+#define UTIL_COLL_BXOR(dst,src)	(dst) ^= (src)
+
+static int util_coll_proc_reduce_item(struct util_coll_mc *coll_mc,
+				      struct util_coll_reduce_item *reduce_item)
+{
+	switch (reduce_item->op) {
+	case FI_MIN:
+		SWITCH_REDUCE(reduce_item->datatype, UTIL_COLL_REDUCE_FUNC,
+			      reduce_item->in_buf, reduce_item->inout_buf,
+			      reduce_item->count, UTIL_COLL_MIN);
+		break;
+	case FI_MAX:
+		SWITCH_REDUCE(reduce_item->datatype, UTIL_COLL_REDUCE_FUNC,
+			      reduce_item->in_buf, reduce_item->inout_buf,
+			      reduce_item->count, UTIL_COLL_MAX);
+		break;
+	case FI_SUM:
+		SWITCH_REDUCE(reduce_item->datatype, UTIL_COLL_REDUCE_FUNC,
+			      reduce_item->in_buf, reduce_item->inout_buf,
+			      reduce_item->count, UTIL_COLL_SUM);
+		break;
+	case FI_PROD:
+		SWITCH_REDUCE(reduce_item->datatype, UTIL_COLL_REDUCE_FUNC,
+			      reduce_item->in_buf, reduce_item->inout_buf,
+			      reduce_item->count, UTIL_COLL_PROD);
+		break;
+	case FI_LOR:
+		SWITCH_REDUCE(reduce_item->datatype, UTIL_COLL_REDUCE_FUNC,
+			      reduce_item->in_buf, reduce_item->inout_buf,
+			      reduce_item->count, UTIL_COLL_LOR);
+		break;
+	case FI_LAND:
+		SWITCH_REDUCE(reduce_item->datatype, UTIL_COLL_REDUCE_FUNC,
+			      reduce_item->in_buf, reduce_item->inout_buf,
+			      reduce_item->count, UTIL_COLL_LAND);
+		break;
+	case FI_BOR:
+		SWITCH_REDUCE(reduce_item->datatype, UTIL_COLL_REDUCE_FUNC,
+			      reduce_item->in_buf, reduce_item->inout_buf,
+			      reduce_item->count, UTIL_COLL_BOR);
+		break;
+	case FI_BAND:
+		SWITCH_REDUCE(reduce_item->datatype, UTIL_COLL_REDUCE_FUNC,
+			      reduce_item->in_buf, reduce_item->inout_buf,
+			      reduce_item->count, UTIL_COLL_BAND);
+		break;
+	case FI_LXOR:
+		SWITCH_REDUCE(reduce_item->datatype, UTIL_COLL_REDUCE_FUNC,
+			      reduce_item->in_buf, reduce_item->inout_buf,
+			      reduce_item->count, UTIL_COLL_LXOR);
+		break;
+	case FI_BXOR:
+		SWITCH_REDUCE(reduce_item->datatype, UTIL_COLL_REDUCE_FUNC,
+			      reduce_item->in_buf, reduce_item->inout_buf,
+			      reduce_item->count, UTIL_COLL_BXOR);
+		break;
+	case FI_ATOMIC_READ:
+	case FI_ATOMIC_WRITE:
+	case FI_CSWAP:
+	case FI_CSWAP_NE:
+	case FI_CSWAP_LE:
+	case FI_CSWAP_LT:
+	case FI_CSWAP_GE:
+	case FI_CSWAP_GT:
+	case FI_MSWAP:
+	case FI_ATOMIC_OP_LAST:
+	case FI_BARRIER:
+	case FI_BROADCAST:
+	case FI_ALLTOALL:
+	case FI_ALLGATHER:
+	default:
+		return -FI_ENOSYS;
+	};
+	return FI_SUCCESS;
+}
```

Comment:
I think all of the above code in this file can basically be replaced by calls to ofi_atomic_xxx_handlers(), or by extending the ofi_atomic_xxx_handlers.